src/control/tools/BaseStrokeHandler.h

```diff
@@ -28,12 +37,24 @@ class BaseStrokeHandler : public InputHandler
 	bool onMotionNotifyEvent(const PositionInputData& pos);
 	void onButtonReleaseEvent(const PositionInputData& pos);
 	void onButtonPressEvent(const PositionInputData& pos);
+	virtual bool onKeyEvent(GdkEventKey* event);
 
 private:
-	virtual void drawShape(Point& currentPoint, bool shiftDown) = 0;
-
+	virtual void drawShape(Point& currentPoint, const PositionInputData& pos) = 0;
+	DIRSET_MODIFIERS drawModifierFixed = NONE;
+	int lastCursor = -1;	//avoid same setCursor
+	bool flipShift = false;	//use to reverse Shift key modifier action. i.e.  for separate Rectangle and Square Tool buttons.
+	bool flipControl = false; //use to reverse Control key modifier action.
+	
 protected:
 	void snapToGrid(double& x, double& y);
+	Point currPoint;
+	/**
+	 * modifyModifiersByDrawDir  -  toggle shift and control modifiers depending on initial drawing direction/
+	 */
+	void modifyModifiersByDrawDir(double width, double height, bool changeCursor = true);
+	bool modShift = false;
+	bool modControl = false;
```

Comment:
Please move the Attributes to the other attributes down, and don't mix it with methods